File: test/artifact/pldi24/splay/splay-bu.c
// 2023, Daan Leijen.
// Sleator and Tarjan's bottom up splay algorithm using parent pointers. (Section 4, page 666)
//
// "Self-Adjusting Binary Search Trees", Daniel Dominic Sleator and Robert Endre Tarjan.
// Journal of the Association for Computing Machinery. Vol. 32, No. 3, July 1985, pp. 652-686.

#define ADDPARENT 1
#include "tree.h"

static tree_t* p(tree_t* x) { return x->parent; }
static tree_t* g(tree_t* x) { return p(p(x)); }

static void rotate_left(tree_t* y) {
  tree_t* x = y->right; tree_t* z = p(y);
  if(z) {
    if (z->left == y) {
      z->left = x;
    } else {
      z->right = x;
    }
  }
  y->right = x->left;
  x->left = y;
  x->parent = z;
  y->parent = x;

  if(y->right) y->right->parent = y;
}

static void rotate_right(tree_t* y) {
  tree_t* x = y->left; tree_t* z = p(y);
  if(z) {
    if (z->right == y) {
      z->right = x;
    } else {
      z->left = x;
    }
  }
  y->left = x->right;
  x->right = y;
  x->parent = z;
  y->parent = x;

  if(y->left) y->left->parent = y;
}

static void splay(tree_t* x) {
  while (p(x)) {
    if (x == p(x)->left) {
      if (g(x) == NULL) { rotate_right(p(x)); }
      else if (p(x) == g(x)->left) { rotate_right(g(x)); rotate_right(p(x)); }
      else { rotate_right(p(x)); rotate_left(p(x)); }
    } else {
      if (g(x) == NULL) { rotate_left(p(x)); }
      else if (p(x) == g(x)->right) { rotate_left(g(x)); rotate_left(p(x)); }
      else { rotate_left(p(x)); rotate_right(p(x)); }
    }
  }
}

static tree_t* access(tree_t* t, tkey_t k) {
  tree_t* p = NULL;
  
  // search
  while (is_node(t)) {
    if (t->key < k) { p = t; t = t->right; }
    else if (t->key > k) { p = t; t = t->left; }
    else break; 
  }
  
  // create if needed
  if (is_leaf(t)) {
    t = node(leaf(),k,leaf());
    t->parent = p;
  }
  
  // link to parent
  if (!p) { return t; }
  else if (p->key < t->key) p->right = t;
  else p->left = t;

  // splay up  
  splay(t);

  return t;
}
  

File: test/artifact/pldi24/splay/splay-bu.hs
import Tree


data Zipper
  = NodeR { left :: !Tree, key :: !Key, up :: !Zipper }
  | NodeL { up :: !Zipper, key :: !Key, right :: !Tree }
  | Done


access :: Tree -> Key -> Tree
access t k  = accessz t k Done

accessz :: Tree -> Key -> Zipper -> Tree
accessz t k z
  = case t of
      Leaf -> splay z Leaf k Leaf
      Node l x r ->
        if   x < k then accessz r k (NodeR l x z)
        else if x > k then accessz l k (NodeL z x r)
        else splay z l x r

splay :: Zipper -> Tree -> Key -> Tree -> Tree
splay z l x r
  = case z of
      Done -> Node l x r
      NodeR zl zx zz -> case zz of
        Done -> Node (Node zl zx l) x r
        NodeR zzl zzx zzz -> splay zzz (Node (Node zzl zzx zl) zx l) x r
        NodeL zzz zzx zzr -> splay zzz (Node zl zx l) x (Node r zzx zzr)
      NodeL zz zx zr  -> case zz of
        Done -> Node l x (Node r zx zr)
        NodeR zzl zzx zzz -> splay zzz (Node zzl zzx l) x (Node r zx zr)
        NodeL zzz zzx zzr -> splay zzz l x (Node r zx (Node zr zzx zzr))

main :: IO ()
main
  = benchMain access
